bin/beepmein

```diff
@@ -0,0 +1,281 @@
+    xrandr --output "$monitor" --brightness "$now"
+    done
+}
+
+# alarm failsafe (by misko), this will be used if 'alarmSample' is not found.
+alarmfailsafe () {
+for ((i=1; i<=10; i++)); do
+      for ((c=1;c<=3;c++)); do
+          timeout -s HUP  0.2 speaker-test -t sine -f 3000  > /dev/null 2>&1
+          sleep 0.2
+       done
+       sleep 1
+done
+}
+
+# test alarm sample path (no test for reminder sample)
+testAlarmPath () {
+    [[ -f "$alarmSample" ]] || { >&2 superecho "Note: alarmSample $alarmSample not found, Internal fail-safe beep will be used instead."; }
+}
+
+# mpv or cvlc (by jr)
+playa () {
+    file="$1"
+    [[ -f $file ]] || { superecho "file not found" ; return 1; }
+    if command -v mpv >/dev/null 2>&1
+    then
+        mpv --no-resume-playback --no-video "$file"
+    elif command -v cvlc >/dev/null 2>&1
+    then
+        cvlc --no-video "$file" vlc://quit
+    else
+        superecho "no player"
+        return 1
+    fi
+}
+
+# test audio and notifications
+test () {
+    echo "notify test"
+    notify-send "notify test" || echo "failed"
+    echo
+    echo "fail-safe internal beep test used for reminders"
+    beep "3" || echo "failed"
+    echo
+    echo "mpv sample test used for alarms (either this or vlc must pass)"
+    mpv --no-resume-playback --no-video --length=10 "$alarmSample" || echo "mpv failed"
+    echo
+    echo "vlc sample test used for alarms"
+    cvlc --no-video --run-time=10 "$alarmSample" vlc://quit || echo "cvlc failed"
+    echo
+    echo "fail-safe internal beep test used for alarms"
+    alarmfailsafe || echo "failed"
+    echo
+}
+
+# validate 'at' time syntax by generating fake job
+validate () {
+tmpjob="$(at "$1" <<< "$(echo "test")" 2>&1 | grep ^job | cut -f2 -d ' ')"
+if [ -n "$tmpjob" ]; then
+    # job was generated ok
+    atrm "$tmpjob"
+else
+    # error
+    return 1
+fi
+}
+
+# help or test
+if [ "$1" == "--help" ] || [ "$1" == "-h" ] || [ $# -eq 0 ]; then
+    help; exit
+elif [ "$1" == "--test" ]; then
+    test; exit
+fi
+
+# list all 'at' jobs, last non-empty line only
+if [ "$1" == "-l" ] || [ "$1" == "--list" ]; then
+    atq | sort -r | while read -r job rest ; do
+        echo -n "$job $rest | "
+        at -c "$job" | tac | grep -vm 1 '^[[:space:]]*$'
+    done
+    exit
+fi
+
+# --set-alarm (via softlink)
+if [ "$1" == "--set-alarm" ]; then
+    if [[ -z "$2" ]]; then
+        alarmpath="$(yad --on-top --center --title="Select alarm sample path" --file)"
+    else
+        alarmpath=$(readlink -f "$2") # must be absolute path
+    fi
+    [[ -f "$alarmpath" ]] || { superecho "No valid alarm path given." ; beep "2"; exit 1; }
+    # f=force
+    ln -fs "$alarmpath" "$config/alarm"
+    exit
+fi
+# --set-remind (via softlink)
+if [ "$1" == "--set-remind" ]; then
+    if [[ -z "$2" ]]; then
+        remindpath="$(yad --on-top --center --title="Select reminder sample path" --file)"
+    else
+        remindpath=$(readlink -f "$2") # must be absolute path
```

Comment:
Important to seperate options from positionals using `--` if the user is allowed to specify input.

Reply:
Example?

Reply:
Just

```
readlink -f -- "$2"
```
Since the user input is a path, `-v` would be a perfectly fine path if it's a relative path (only absolute paths are guaranteed to start with a /). So assuming that "$2" may be an arbitrary relative path, if the path is `-v`, the command would expand to `readlink -f -v` which will produce a failure since readlink will assume that -v is an option (a valid one in this case) and expect another argument. Inserting `--` is a convention implemented by many argument parsers, and in all (if not most) GNU utilities, that signals the end of options and makes everything later into a positional argument that will never be interpreted as an option. In our case, `-v` would then be a valid positional argument and not an option name.